connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -813,6 +823,28 @@ class SparkSession private[sql] (
    * Set to false to prevent client.releaseSession on close() (testing only)
    */
   private[sql] var releaseSessionOnClose = true
+
+  private[sql] def registerObservation(planId: Long, observation: Observation): Unit = {
+    // makes this class thread-safe:
+    // only the first thread entering this block can set sparkSession
+    // all other threads will see the exception, as it is only allowed to do this once
+    observation.synchronized {
+      if (observationRegistry.contains(planId)) {
+        throw new IllegalArgumentException("An Observation can be used with a Dataset only once")
+      }
+      observationRegistry.put(planId, observation)
+    }
+  }
+
+  private[sql] def setMetricsAndUnregisterObservation(
+      planId: Long,
+      metrics: Option[Map[String, Any]]): Unit = {
+    observationRegistry.get(planId).map { observation =>
+      if (observation.setMetricsAndNotify(metrics)) {
+        observationRegistry.remove(planId)
```

Comment:
Should this be tied to whether or not the observation has been successfully updated? Other question under what circumstance can the metrics be empty.

Reply:
I had the same question when I looked at the code. In Spark Core we only de-register the Observation when some non-empty metrics are set. I am not sure under which circumstance the metrics can be empty.

Reply:
I looked at the code. It seems it's valid to check for non-empty metrics in Spark Core:
```scala
private[spark] def onFinish(qe: QueryExecution): Unit = {
  ...
  val row: Option[Row] = qe.observedMetrics.get(name)
  val metrics: Option[Map[String, Any]] = row.map(r => r.getValuesMap[Any](r.schema.fieldNames.toImmutableArraySeq))
  if (setMetricsAndNotify(metrics)) {
    unregister()
  }
}
```

The option is for a case when the query finishes without metric. Is this possible?

Reply:
But nevertheless we don't need to handle this case in Connect because we look up using Plan Id.